File: src/Components/HmacResultFactory.cs
namespace HmacManager.Components;

/// <summary>
/// Factory class for creating <see cref="HmacResult"/> instances.
/// </summary>
internal class HmacResultFactory : IHmacResultFactory
{
    /// <summary>
    /// Gets the policy associated with the HMAC result.
    /// </summary>
    protected readonly string Policy;

    /// <summary>
    /// Gets the optional header scheme associated with the HMAC result.
    /// </summary>
    protected readonly string? HeaderScheme;

    /// <summary>
    /// Initializes a new instance of the <see cref="HmacResultFactory"/> class with the specified policy and optional header scheme.
    /// </summary>
    /// <param name="policy">The policy to associate with the HMAC result.</param>
    /// <param name="headerScheme">The optional header scheme to associate with the HMAC result.</param>
    internal HmacResultFactory(string policy, string? headerScheme = null)
    {
        Policy = policy;
        HeaderScheme = headerScheme;
    }

    /// <summary>
    /// Creates a successful <see cref="HmacResult"/> with the specified <see cref="Hmac"/> data.
    /// </summary>
    /// <param name="hmac">The HMAC data to include in the result.</param>
    /// <returns>A successful <see cref="HmacResult"/> instance.</returns>
    public HmacResult Success(Hmac hmac) => Create(isSuccess: true, hmac);

    /// <summary>
    /// Creates a failed <see cref="HmacResult"/>.
    /// </summary>
    /// <returns>A failed <see cref="HmacResult"/> instance.</returns>
    public HmacResult Failure() => Create(isSuccess: false);

    /// <summary>
    /// Creates an <see cref="HmacResult"/> with the specified success status and optional HMAC data.
    /// </summary>
    /// <param name="isSuccess">A boolean value indicating whether the result represents success.</param>
    /// <param name="hmac">The optional HMAC data to include in the result.</param>
    /// <returns>An <see cref="HmacResult"/> instance.</returns>
    private HmacResult Create(bool isSuccess, Hmac? hmac = null) =>
        new HmacResult(isSuccess, hmac);
}

File: client/lib/hmac_manager/components/hmac-result-factory.ts
import { HmacResult } from "./hmac-result.js";
import { Hmac } from "./hmac.js";

/**
 * Factory class for creating HMAC result instances.
 */
export class HmacResultFactory {
    /**
     * Creates a successful HMAC result.
     * @param hmac - The HMAC details to include in the result.
     * @returns An HMAC result indicating success.
     */
    success = (hmac: Hmac) => this.create(true, hmac);

    /**
     * Creates a failed HMAC result.
     * @returns An HMAC result indicating failure.
     */
    failure = () => this.create(false);

    /**
     * Creates an HMAC result instance.
     * @param isSuccess - Indicates whether the result is successful.
     * @param hmac - The HMAC details, or null if the result is a failure.
     * @returns An HMAC result object.
     */
    private create = (isSuccess: boolean, hmac: Hmac | null = null): HmacResult => ({
        hmac,
        isSuccess,
        dateGenerated: new Date()
    });
}